Use latexmk for pdf (and dvi) targets

jfbu opened the pull request:
This will need some review because I don't know if latexmk is available on all platforms where Sphinx is available. I know latexmk comes with TeXLive installations and I think the TeXLive has its own perl so that latexmk runs (also on windows).

rel #3064, #2513

```
modified:   sphinx/texinputs/Makefile_t
new file:   sphinx/texinputs/latexmkrc
```

jfbu committed:
Use latexmk for pdf (and dvi) targets (on unixen)

jfbu commented:
latexmk is a Perl script, and the necessary Perl comes with standard TeXLive installs. On MS-Windows, with MikTeX, it is needed that Perl be separately installed. 

I tested with TeXLive2013 on compiling the Sphinx own docs to pdf, and it worked with no issue.

note: this was with TL2013 not the initial TL2013, it had latexmk 10 Nov 2013. Version 4.39 --- I have now also tested with final TL2012 and its latexmk 11 Nov. 2012, version: 4.35 and it seems to run fine.

Japanese documents would need separate testing ... (based on advice from latexmk's author, all errors are my own additions...).

**side note** I have tested with compiling Sphinx own pdf docs under pretending it was with language ``'ja'`` and it seems ok, and actually  I am jealous of the look of the index which is much nicer from `make latexpdfja` than from `make latexpdf` (use of a smaller font size, a column separator...)

edit: there was indeed a silly error of my own in the `latexmkjarc` config file I set-up for `latexmk` with usage of `platex`, and I have fixed it in forced update of commit 2.

I tried to set-up a working template for both pdf and dvi output. I have added support for glossaries (in the LaTeX sense, with `glo` and `gls` files) but I have no idea if Sphinx currently supports this (the original template for `make latexpdf`  contained no related set-up).

Setting the milestone to 1.6 ...

jfbu commented:
Last commit is analog of PR #3322 in this context with latexmk. There is also a `-silent` option for `latexmk` but it silences all sub-processes, hence appears too drastic.

Here is for example bottom of console output from `make latexpdf` of the (1.5a3) Sphinx docs, with `latex_engine` set to `'xelatex'`.

```
...
copying TeX support files...
done
build succeeded, 1 warning.
make[1] : on entre dans le répertoire « /home/FOO/sphinx/doc/_build/latex »
latexmk -pdflatex="xelatex -halt-on-error -interaction=batchmode %O %S" -pdf -dvi- -ps-  'sphinx.tex'
Latexmk: This is Latexmk, John Collins, 5 Sep. 2016, version: 4.48.
Latexmk: applying rule 'pdflatex'...
Rule 'pdflatex': Rules & subrules not known to be previously run:
   pdflatex
Rule 'pdflatex': The following rules & subrules became out-of-date:
      'pdflatex'
------------
Run number 1 of rule 'pdflatex'
------------
------------
Running 'xelatex -halt-on-error -interaction=batchmode  -recorder  "sphinx.tex"'
------------
This is XeTeX, Version 3.14159265-2.6-0.99996 (TeX Live 2016) (preloaded format=xelatex)
 restricted \write18 enabled.
entering extended mode
Latexmk: Index file 'sphinx.idx' was written
Latexmk: References changed.
Latexmk: Missing input file: 'sphinx.ind' from line
  'No file sphinx.ind.'
Latexmk: References changed.
Latexmk: Log file says output to 'sphinx.pdf'
Latexmk: List of undefined refs and citations:
  Reference `builders:module-sphinx.builders' on page 289 undefined on input line 25277
  Reference `builders:module-sphinx.builders.applehelp' on page 289 undefined on input line 25278
  Reference `builders:module-sphinx.builders.changes' on page 289 undefined on input line 25279
  Reference `builders:module-sphinx.builders.devhelp' on page 289 undefined on input line 25280
  Reference `builders:module-sphinx.builders.dummy' on page 289 undefined on input line 25281
  Reference `builders:module-sphinx.builders.epub' on page 289 undefined on input line 25282
  Reference `builders:module-sphinx.builders.epub3' on page 289 undefined on input line 25283
  Reference `builders:module-sphinx.builders.gettext' on page 289 undefined on input line 25284
  Reference `builders:module-sphinx.builders.html' on page 289 undefined on input line 25285
  Reference `builders:module-sphinx.builders.htmlhelp' on page 289 undefined on input line 25286
  Reference `builders:module-sphinx.builders.latex' on page 289 undefined on input line 25287
  Reference `builders:module-sphinx.builders.linkcheck' on page 289 undefined on input line 25288
  Reference `builders:module-sphinx.builders.manpage' on page 289 undefined on input line 25289
  Reference `builders:module-sphinx.builders.qthelp' on page 289 undefined on input line 25290
  Reference `builders:module-sphinx.builders.texinfo' on page 289 undefined on input line 25291
  Reference `builders:module-sphinx.builders.text' on page 289 undefined on input line 25292
  Reference `builders:module-sphinx.builders.xml' on page 289 undefined on input line 25293
  Reference `config:module-conf' on page 289 undefined on input line 25296
  Reference `ext/autodoc:module-sphinx.ext.autodoc' on page 289 undefined on input line 25306
  Reference `ext/autosectionlabel:module-sphinx.ext.autosectionlabel' on page 289 undefined on input line 25307
  Reference `ext/autosummary:module-sphinx.ext.autosummary' on page 289 undefined on input line 25308
  Reference `ext/coverage:module-sphinx.ext.coverage' on page 289 undefined on input line 25309
  Reference `ext/doctest:module-sphinx.ext.doctest' on page 289 undefined on input line 25310
  Reference `ext/extlinks:module-sphinx.ext.extlinks' on page 289 undefined on input line 25311
  Reference `ext/githubpages:module-sphinx.ext.githubpages' on page 289 undefined on input line 25312
  Reference `ext/graphviz:module-sphinx.ext.graphviz' on page 289 undefined on input line 25313
  Reference `ext/ifconfig:module-sphinx.ext.ifconfig' on page 289 undefined on input line 25314
  Reference `ext/inheritance:module-sphinx.ext.inheritance_diagram' on page 289 undefined on input line 25316
  Reference `ext/intersphinx:module-sphinx.ext.intersphinx' on page 289 undefined on input line 25317
  Reference `ext/linkcode:module-sphinx.ext.linkcode' on page 289 undefined on input line 25319
  Reference `ext/math:module-sphinx.ext.imgmath' on page 289 undefined on input line 25315
  Reference `ext/math:module-sphinx.ext.jsmath' on page 289 undefined on input line 25318
  Reference `ext/math:module-sphinx.ext.mathbase' on page 289 undefined on input line 25320
  Reference `ext/math:module-sphinx.ext.mathjax' on page 289 undefined on input line 25321
  Reference `ext/napoleon:module-sphinx.ext.napoleon' on page 289 undefined on input line 25322
  Reference `ext/todo:module-sphinx.ext.todo' on page 289 undefined on input line 25323
  Reference `ext/viewcode:module-sphinx.ext.viewcode' on page 289 undefined on input line 25324
  Reference `extdev/appapi:module-sphinx.application' on page 289 undefined on input line 25274
  Reference `extdev/appapi:module-sphinx.config' on page 289 undefined on input line 25297
  Reference `extdev/appapi:module-sphinx.errors' on page 289 undefined on input line 25305
  Reference `extdev/domainapi:module-sphinx.domains' on page 289 undefined on input line 25301
  Reference `extdev/envapi:module-sphinx.environment' on page 289 undefined on input line 25304
  Reference `extdev/markupapi:module-docutils.parsers.rst' on page 289 undefined on input line 25300
  Reference `extdev/nodes:module-sphinx.addnodes' on page 289 undefined on input line 25273
  Reference `extdev/parserapi:module-sphinx.parsers' on page 289 undefined on input line 25330
  Reference `latex:module-latex' on page 289 undefined on input line 25327
Latexmk: Summary of warnings:
  Latex failed to resolve 46 reference(s)
Latexmk: applying rule 'makeindex sphinx.idx'...
Rule 'makeindex sphinx.idx': File changes, etc:
   Non-existent destination files:
      'sphinx.ind'
------------
Run number 1 of rule 'makeindex sphinx.idx'
------------
------------
Running 'makeindex -s python.ist  -o "sphinx.ind" "sphinx.idx"'
------------
This is makeindex, version 2.15 [TeX Live 2016] (kpathsea + Thai support).
Scanning style file ./python.ist......done (6 attributes redefined, 0 ignored).
Scanning input file sphinx.idx.....done (1164 entries accepted, 0 rejected).
Sorting entries............done (12539 comparisons).
Generating output file sphinx.ind.....done (1571 lines written, 0 warnings).
Output written in sphinx.ind.
Transcript written in sphinx.ilg.
Latexmk: applying rule 'pdflatex'...
Rule 'pdflatex': File changes, etc:
   Changed files, or newly in use since previous run(s):
      'sphinx.aux'
      'sphinx.ind'
------------
Run number 2 of rule 'pdflatex'
------------
------------
Running 'xelatex -halt-on-error -interaction=batchmode  -recorder  "sphinx.tex"'
------------
This is XeTeX, Version 3.14159265-2.6-0.99996 (TeX Live 2016) (preloaded format=xelatex)
 restricted \write18 enabled.
entering extended mode
Latexmk: Index file 'sphinx.idx' was written
Latexmk: Log file says output to 'sphinx.pdf'
Latexmk: applying rule 'pdflatex'...
Rule 'pdflatex': File changes, etc:
   Changed files, or newly in use since previous run(s):
      'sphinx.aux'
      'sphinx.out'
      'sphinx.toc'
------------
Run number 3 of rule 'pdflatex'
------------
------------
Running 'xelatex -halt-on-error -interaction=batchmode  -recorder  "sphinx.tex"'
------------
This is XeTeX, Version 3.14159265-2.6-0.99996 (TeX Live 2016) (preloaded format=xelatex)
 restricted \write18 enabled.
entering extended mode
Latexmk: Index file 'sphinx.idx' was written
Latexmk: Log file says output to 'sphinx.pdf'
Latexmk: All targets (sphinx.pdf) are up-to-date
make[1] : on quitte le répertoire « /home/FOO/sphinx/doc/_build/latex »
```

jfbu commented:
As with #3322 I am hoping that ``-halt-on-error -interaction=batchmode`` does work as command line options for the MikTeX `*tex` executables (I don't have a MS-Windows box).

edit: according to https://docs.miktex.org/manual/pdftex.html (also [xetex](https://docs.miktex.org/manual/xetex.html) and I presume `luatex` too), they do, at least in the form with two hyphens:

    --halt-on-error --interaction=batchmode

and fortunately this syntax is also accepted by the TeXLive `*tex` executables.

Regarding `Latexmk` here are <strike>its</strike> some [docs](http://mg.readthedocs.io/latexmk.html). **edit**: oh sorry, this is but a useful summary, not the official documentation. Apologies for confusion. Never trust first link of startpage.com's output ;-)

Home page of `Latexmk` is [here](http://www.personal.psu.edu/jcc8/latexmk/).

tk0miya commented:
As a trial, I failed to build japanese docs with latexmk.
```
$ sphinx-quickstart test
$ cd test
$ echo "language = 'ja'" >> conf.py
$ make latexpdf
...
------------
Run number 1 of rule 'makeindex test.idx'
------------
------------
Running 'mendex -U -f -d "test".dic -s python.ist  -o "test.ind" "test.idx"'
------------
This is mendex version 2.6f [14-Aug-2009] (utf8.uptex) (TeX Live 2016).
Warning: Couldn't find dictionary file test.dic.
Scanning input file test.idx....done (0 entries accepted, 0 rejected).
0 entries accepted, 0 rejected.
Nothing written in output file.
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  makeindex test.idx: Command for 'makeindex test.idx' gave return code 65280
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
make[1]: *** [all-pdf-ja] Error 12
```

I could not understand what the error means. So I will ask japanese TeX experts about the error later.
Anyway, before merging, we have to confirm this works fine on windows.

@shimizukawa @usaturn could you confirm this on windows platfrom please?

jfbu commented:
<strike>@tk0miya could you try with ``make latexpdfja`` rather ? The ``makeindex`` was invoked but it should have been ``mendex``. Does one compile Japanese docs with ``make latexpdf``? I thought it was with ``make latexpdfja``.</strike>

jfbu commented:
ah sorry, please do not consider my previous comment. I got fooled by console output.

tk0miya commented:
I also tried `make latexpdfja`, but I got same error. It seems `mendex` command was invoked correctly.

jfbu commented:
I am investigating. The return code 65280 is to be understood as 65280/256=255. It is a return code from `mendex`. Perhaps it is benign exit code, but latexmk considers it as signaling error.

jfbu commented:
It seems to be only artefact from an empty `test.idx` file. If I add

```
.. py:function:: enumerate(sequence[, start=0])

   Return an iterator that yields tuples of an index and an item of the
   *sequence*. (And so on.)
```

for example to `index.rst` then all works fine.

I looked quickly at the man page for `mendex` in TL2016 (it is in English) but it has no info on exit codes. I will write to `latexmk` author to ask how to handle such return code `255` of mendex in such case.

As per Windows, perhaps the hyphens need to be doubled:

[mypatch.txt](https://github.com/sphinx-doc/sphinx/files/705992/mypatch.txt)

jfbu commented:
I will push commit with doubled hyphens because syntax works on TeXLive and is the documented one for MikTeX.

jfbu commented:
This really seems to be it. I tested with
```latex
\documentclass{jsbook}
% This file testmendex.tex
% Try:
% platex testmendex.tex
% Then:
% mendex testmendex.idx
% exit code 255 if empty idx. And no ind file, even a zero byte one.
% exit code 0 if with \index line below de-commented
\makeindex
\begin{document}
Hello. I don't speak Japanese!
%\index{whatever}
\end{document}
```
<strike>Earlier it went unnoticed possibly because the mendex call was followed in Makefile by platex calls.</strike> The Makefile recipe line doing `mendex` was prefixed with `-` hence the error 255 was reported but ignored.

I don't know if Latexmk stops because of non-zero exit code or because `.ind` file is not produced. Notice that in similar situation `makeindex` produces a zero byte `.ind` file and has exit code zero, but `mendex` produces no file at all and has exit code 255.

jfbu commented:
One could use

    $makeindex = 'mendex -U -f -d %B.dic -s python.ist %S || touch %B.ind';

in `sphinx/texinputs/latexmkjarc`. But my Windows ignorance makes that I don't know if this would work there. Errors encountered by `mendex` would still be reported to user.

update: everybody but me knows the `||` work on Windows but the problem is with `touch`. Perhaps something with `echo` then in place of `touch`?

tk0miya commented:
LGTM. `||` works fine on MacOS.

About windows, let's see reports from Japanese developers who use windows.

jfbu commented:
I got some advice from Latexmk's author. He mentioned that in the current ``all-pdf-ja`` rule from ``<build>/latex/Makefile`` which I showed him the mendex invocation (with the ``-`` prefix to ignore the error code) has the problem that if for whatever reason there is a corrupted ``foo.ind`` pre-existing file, and the ``foo.tex`` generates an empty ``foo.idx``, then the mendex call does not erase the previously existing file, which is then used as is for final ``foo.dvi``. But as in Sphinx workflow the ``<build>`` directory is only  a work-directory, possibly this defect did not cause issues so far.

Taking this remark into account, regarding the Perl rule in `latexmkjarc`, a better choice will be

    $makeindex = 'rm %D; mendex -U -f -d %B.dic -s python.ist %S || touch %D';

where ``%D`` is the Latexmk place-holder for the target (``foo.ind``). J. C. recommended ``%D`` as it takes into account ``-output-directory`` option which may be passed to ``*tex`` executables.

As the above does not work as is on Windows, he suggests a Perl solution would be better in long run and gave me some advice about where to start.

With unix-like syntax I feel this could be good:

    $makeindex = 'rm -f %D; mendex -U -f -d %B.dic -s python.ist %S || echo "mendex exited with error code $? (ignoring and touching %D)" && touch %D';

I tested it on mac os:
```bash
[...]
Run number 1 of rule 'makeindex test.idx'
------------
------------
Running 'rm -f "test.ind"; mendex -U -f -d "test".dic -s python.ist "test.idx" || echo "mendex exited with error code $? (ignoring and touching "test.ind")" && touch "test.ind"'
------------
This is mendex version 2.6f [14-Aug-2009] (utf8.uptex) (TeX Live 2016).
Warning: Couldn't find dictionary file test.dic.
Scanning input file test.idx....done (0 entries accepted, 0 rejected).
0 entries accepted, 0 rejected.
Nothing written in output file.
mendex exited with error code 255 (ignoring and touching test.ind)
Latexmk: applying rule 'latex'...
Rule 'latex': File changes, etc:
   Changed files, or newly in use since previous run(s):
      'test.aux'
      'test.ind'
------------
Run number 2 of rule 'latex'
[...]
```

Windows syntax will be something else.

tk0miya commented:
ping @shimizukawa @usaturn 
>could you confirm this on windows platfrom please?

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -7,34 +7,35 @@ ALLPS  = $(addsuffix .ps,$(ALLDOCS))
 # Prefix for archive names
 ARCHIVEPRREFIX =
-# Additional LaTeX options
+# Additional LaTeX options.
```
sorry about added dot at the end of this comment, my mistake

jfbu replied:
and I forgot to remove it again ...

shimizukawa commented on sphinx/texinputs/latexmkjarc:
```diff
@@ -0,0 +1,7 @@
+$latex = 'platex --halt-on-error --interaction=nonstopmode -kanji=utf8 %O %S';
+$dvipdf = 'dvipdfmx %O -o %D %S';
+$makeindex = 'rm -f %D; mendex -U -f -d %B.dic -s python.ist %S || echo "mendex exited with error code $? (ignoring)" && : >> %D';
```
Windows doesn't have `rm` command, but `del`.
Is there any workaround?
(commented out this line works well for small documentation)

tk0miya replied:
I don't know better solution, but, as a workaround, we can provide another rc file (cf. `latexmkrc.win`).

jfbu replied:
I did the `rm` based on a comment by Latexmk author:

> Actually, there's one problem: if there's a previously existing .ind file, it will not be deleted.  In that case you end up with an .ind file with wrong information in it instead of a zero length file..

Thus it seems we are considering here a situation where
- we have some ind file handing around
- the tex file in the end contains no `\index` command hence would generate non ind file by itself

Thus perhaps for simplicity sake it is better not to worry. It should be rare when suddenly all `\index` have disappeared. Hence let's just drop this (if we can find a working solution to other things).

jfbu replied:
In the end, it was decided to keep usage to unix platforms, so the `rm -f %D` was re-added ...

shimizukawa commented on sphinx/texinputs/Makefile_t:
```diff
@@ -57,30 +58,18 @@ bz2: tar
-	-$(MAKEINDEX) -s python.ist '$(basename $<).idx'
-	$(PDFLATEX) $(LATEXOPTS) '$<'
+%.pdf: %.tex FORCE_MAKE
 	$(PDFLATEX) $(LATEXOPTS) '$<'
```
On windows, latexmk recognize as `'` quote string is a part of filename. Then it will cause following error:
```
Latexmk: Could not find file ['3082.tex'].
```
When I replace `'$<'` in the Makefile with `"$<"`, it works fine.

On Windows:
```
>python - "Foo Bar" 'Foo Bar'
Python 3.5.2 |Anaconda 4.2.0 (64-bit)| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; sys.argv
['-', 'Foo Bar', "'Foo", "Bar'"]
```

On Linux:
```
>docker run -it --rm python:3.6 bash
root@7d726b25ca1b:/# python - "Foo Bar" 'Foo Bar'
Python 3.6.0 (default, Jan 18 2017, 02:51:38)
[GCC 4.9.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; sys.argv
['-', 'Foo Bar', 'Foo Bar']
```

jfbu replied:
This is very useful to me thanks !

I mention in passing that Latexmk by default uses `"` quotes around strings from its place holders like `%O`, `%S`. Thus on Linux we must be carefult not to had one more  `"` layer. But this is something else than the Make variables.

shimizukawa commented on sphinx/texinputs/Makefile_t:
```diff
@@ -7,34 +7,35 @@ ALLPS  = $(addsuffix .ps,$(ALLDOCS))
 all-ps: $(ALLPS)
 
 all-pdf-ja:
 	for f in *.pdf *.png *.gif *.jpg *.jpeg; do extractbb $$f; done
```
I think `all-pdf-ja` is not depend to "Japanese", so I suggest as following:
```
ALLIMGS = $(wildcard *.png *.gif *.jpg *.jpeg)

{% if latex_engine == 'platex' %}
LATEX = latexmk -r latexmkjarc -pdfdvi -dvi- -ps-
PDFLATEX = $(LATEX)

all: all-dvi $(ALLPDF)
all-pdf: all-dvi $(ALLPDF)
{% else %}
LATEX = latexmk -dvi
PDFLATEXCMD = {{ latex_engine }} --halt-on-error --interaction=nonstopmode
PDFLATEX = latexmk -pdflatex="$(PDFLATEXCMD) %%O %%S" -pdf -dvi- -ps-

all: $(ALLPDF)
all-pdf: $(ALLPDF)
{% endif -%}

{% if latex_engine == 'lualatex' %}
LATEXOPTS += -lualatex
{% endif -%}

all-dvi: $(ALLIMGS) $(ALLDVI)
all-ps: $(ALLPS)

%.png %.gif %.jpg %.jpeg:
	extractbb $@

#<-- snip -->

{% if latex_engine == 'platex' -%}
%.pdf: %.dvi FORCE_MAKE
{%- else -%}
%.pdf: %.tex FORCE_MAKE
{%- endif %}
	$(PDFLATEX) $(LATEXOPTS) "$<"
```
(Note that I'm not familiar to Makefile)

jfbu replied:
This looks very good idea indeed. Memo: since TeXLive 2015 dvipdfmx does not need the extractbb runs. But it needs them for earlier releases.

shimizukawa commented on sphinx/texinputs/Makefile_t:
```diff
@@ -7,34 +7,35 @@ ALLPS  = $(addsuffix .ps,$(ALLDOCS))
-MAKEINDEX = makeindex
+LATEX = latexmk -dvi
+PDFLATEXCMD = {{ latex_engine }} --halt-on-error --interaction=nonstopmode
+PDFLATEX = latexmk -pdflatex="$(PDFLATEXCMD) %O %S" -pdf -dvi- -ps-
```
I think `%O` and `%S` must be `%%O` and `%%S`.
On my environment (Windows + gmake for win32) cause error as:
```
perl.exe c:\...\latexmk.pl -pdflatex="pdflatex --halt-on-error --interaction=nonstopmode S" -pdf -dvi- -ps-  "3082.tex"
```
`...=nonstopmode S"`: S is not intended.

jfbu replied:
unfortunately ``%%O`` and ``%%S`` cause error on mac os x.

I tried various things, based on Latexmk manual, for example using the `-e` option, but I have always have the problem of diverging behaviour of quoting. I can get things working out on mac os x but I think they would not on Windows.

As a result I must find a way to put all into `latexmkrc` file which is handled by Perl.

shimizukawa replied:
I guess it is a differences of BSD make and GNU make ... :(

shimizukawa reviewed (commented):
@jfbu Thanks for your Makefile improvement with using latexmk! I make some comment for about Windows and platex (ja) mode.

shimizukawa commented:
Full version of Makefile_t that I modified.
[latexmk-makefile-shimizukawa.zip](https://github.com/sphinx-doc/sphinx/files/802018/latexmk-makefile-shimizukawa.zip)

tk0miya commented:
Note: It would be nice if we provide .bat file for windows environment.
@shimizukawa could you make it after this merged?

jfbu commented:
Thanks @shimizukawa for Makefile_t  I am going to use it to replace current one.

As said above I must find a way to put everything in `latexmkrc` file to avoid shell quoting problems. Right now I see a fall-back possibility which is to use ``latex_engine`` in the template which ``latexmkrc`` file to use, if I am not sure whether ``latexmkrc`` can use shell environment variable from Makefile; I will for sure know soon on mac os x, but I have no windows box. Well perhaps I will go for the multiple ``latexmkrc`` files.

By the way, I don't know offhand how to decide from Makefile if running under Unixes or Windows... sorry for ignorance... (but it may not be needed if method above works with Perl syntax only)

jfbu commented:
I have rebased on master and incorporated @shimizukawa Makefile ideas for images.

In the end I only needed one `latexmkrc` file (silly me).

In passing, `xelatex` is faster (when large graphics files are in the document), because intermediate runs produce an `xdv` file, and Latexmk then does the final `xdvipdfmx`.

I hope this works on Windows. As explained in commit log I removed the `rm -f %D;` needed by preparation for `mendex` rule. Only reason is because syntax on windows is with `del`.

Does it work on Windows ?

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -15,38 +21,33 @@ FMT = pdf
+{%- endif %}
 
 %.png %.gif %.jpg %.jpeg:
 	extractbb $@
```
@shimizukawa I incorporated from your Makefile but I see a problem here. For Japanese documents dvipdfmx is able to integrate pdf images. This is I presume the reason original Makefile had a for loop including all pdf files (not very satisfying). Does pdf image inclusion with dvipdfmx not need extractbb ?
By the way, it seems dvipdfmx since TeXLive 2015 but not earlier does the extractbb calls itself. (the extractbb binary is listed under safe executables which don't need shell-escape). Thus to test the above, one needs older installation.

shimizukawa replied:
Nice catch.. 
I don't know whether `extractbb *.pdf` is needed or not.
In fact, In my environment was crushed if `%.pdf` is specified.

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -82,11 +87,14 @@ xz: tar
 
 clean:
-	rm -f *.log *.ind *.aux *.toc *.syn *.idx *.out *.ilg *.pla *.ps *.tar *.tar.gz *.tar.bz2 *.tar.xz $(ALLPDF) $(ALLDVI) *.fls *.fdb_latexmk
+	rm -f *.log *.ind *.aux *.toc *.syn *.idx *.out *.ilg *.pla *.ps *.tar *.tar.gz *.tar.bz2 *.tar.xz $(ALLPDF) $(ALLDVI) $(ALLXDV) *.fls *.fdb_latexmk
```
I am much confused here. This clean rule will not work on Windows right ?

I must admit I have absolutely no idea how Makefiles run on Windows. I see mention on the net of things like MSYS and MinGW, but what kind of shell is that ?

Currently I removed a `rm -f %D;` in latexmkjarc because the command line needs `del` on Windows.

Perhaps Makefile enablers on Windows translate a `rm`  into `del` so the Makefile has no such restriction while the latexmk config file latexmkrc however does have. Lost in the agonies of ignorance ...

shimizukawa replied:
Actually, I invoke `make clean` in `_build/latex`, I get this result:
```
(venv) C:\Project\sphinx-dev\issues\3082\_build\latex>make clean
rm -f *.log *.ind *.aux *.toc *.syn *.idx *.out *.ilg *.pla *.ps *.tar *.tar.gz *.tar.bz2 *.tar.xz 3082.pdf 3082.dvi *.fls *.fdb_latexmk
process_begin: CreateProcess(NULL, rm -f *.log *.ind *.aux *.toc *.syn *.idx *.out *.ilg *.pla *.ps *.tar *.tar.gz *.tar.bz2 *.tar.xz 3082.pdf 3082.dvi *.fls *.fdb_latexmk, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [setup] Error 2
```
I think this is a different new 'gmake on windows' problem...

jfbu reviewed (commented):
@shimizukawa can you check that please ?

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -57,30 +73,28 @@ bz2: tar
 %.ps: %.dvi
-	dvips '$<'
+	dvips "$<"
+{%- endif %}
```
- the quotes here have been changed due to Windows.
- it would be possible to provided ps target via ps2pdf from pdf

jfbu reviewed (commented):
Globally I removed dvi targets if latex_engine is xelatex. Maybe this is wrong. Maybe latex_engine shoud influence only PDF target. But then Makefile would become more complicated. Perhaps it is not needed to go that way. In fact as latex_engine currently does not include `latex+dvips+ps2pdf` route, I wonder why we keep ps target.

shimizukawa commented on sphinx/texinputs/latexmkjarc:
```diff
@@ -0,0 +1,7 @@
+$latex = 'platex --halt-on-error --interaction=nonstopmode -kanji=utf8 %O %S';
+$dvipdf = 'dvipdfmx %O -o %D %S';
+$makeindex = 'mendex -U -f -d %B.dic -s python.ist %S || echo "mendex exited with error code $? (ignoring)" && : >> %D';
```
This line cause a error on Windows as:
```
------------
Running 'mendex -U -f -d "3082".dic -s python.ist "3082.idx" || echo "mendex exited with error code $? (ignoring)" && : >> "3082.ind"'
------------
This is mendex version 2.6f [14-Aug-2009] (utf8.uptex) (TeX Live 2016/W32TeX).
Warning: Couldn't find dictionary file 3082.dic.
Scanning input file 3082.idx....done (0 entries accepted, 0 rejected).
0 entries accepted, 0 rejected.
Nothing written in output file.
"mendex exited with error code $? (ignoring)"
':' is not recognized as an internal or external command, operable program or batch file.
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  makeindex 3082.idx: Command for 'makeindex 3082.idx' gave return code 256
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
C:\Develop\texlive\2016\bin\win32\runscript.tlu:679: command failed with exit code 12:
perl.exe c:\Develop\texlive\2016\texmf-dist\scripts\latexmk\latexmk.pl -r latexmkjarc -dvi  "3082.tex"
make: *** [3082.dvi] Error 12
```

I'm searching a solution ...

shimizukawa replied:
On windows, `type NUL >> %D` works fine. But other platform don't.
One possibility is to run a perl script to change the timestamp.

jfbu replied:
@shimizukawa thanks for this info, I guess I have to plunge into the Perl examples coming with Latexmk to see how it handles things. Also as suggested by @tk0miya we can have per platform latexmkrc files if needed, but then the problem is I don't know offhand how to check from the Makefile the platform... (I am lacking basic software knowledge !)

tk0miya replied:
My proposal is ...
* `Makefile` and `latexmkrc` is used for Unix like OS
* `make.bat` and `latexmkrc.bat` is used for Windows platform

In windows, `make.bat` on current directory is preferentially used than `make.exe`. So providing `make.bat` helps to switch latexmkrc files by platform.

jfbu commented:
I have decided to
- re-configure Makefile and latexmkjarc for unix-like platform (use of ``'``, not ``"``, re-use the ``rm`` needed preliminary to ``mendex``...)
- re-use a shell ``for`` loop  for extractbb and pdf images, in the case of ``dvi`` targets however I may have introduced problem now for platex engine and ``pdf`` target, (I simplified too much logic)
- simplify ``xelatex`` by dropping all complications around ``xdv`` files

jfbu commented:
I have now squashed and rebased. Sorry that previous discussion is now on outdated commits. Here is description:

- use Latexmk as it speeds up compilation to PDF target considerably for big documents, via a reduced number of runs, and it knows how to adapt to more complex situations out of the box (production of glossaries, for example),
- the Makefile is for Unix like platforms (Mac OS X and GNU/Linux), as well as the syntax for shell commands in the latexmkrc files,
- the `xelatex` engine is faster than earlier (even for the same number of runs) for documents having many large graphics, because intermediate runs are with `xdv` intermediate format (a format used by `xelatex` as intermediate step) and only final run produces pdf,
- the `platex` engine triggers use of `mendex` which needs special wrapping because it complains in case of empty ind  file; formerly this was handled by a `-` prefix in GNU Makefile, now this happens from a Latexmk rule hence is done otherwise,
- the `platex` which uses `dvipdfmx` needs bounding box information via `extractbb` for graphics inclusion: this is partially handled via Makefile rule, but images with `pdf` format are handled via a shell for loop (all images were formerly treated via shell for loop, and @shimizukawa introduced idea of using Makefile rules; there remained the case of `pdf` images but it appeared complicated to me to distinguish image file with pdf extension from pdf target files with also pdf extension, from Makefile point of view); as `pdf` target now works without going through `dvi` rule, these `extractbb` calls are added as prequisites to `pdf` target too. By the way, on TeXLive 2015 and later, it is not needed to do `extractbb`, because `dvipdfmx` driver for `graphicx` does it itself (without needing `--shell-escape` flag for TeX run)

Thanks @shimizukawa and @tk0miya for suggestions and comments so far! In the
end I dropped idea to achieve multi-platform Makefile as I lack the needed
competence and can't test, hence I re-incorporated the Unix-only idiom. Sadly,
for same reasons, I don't have knowledge to translate this into ``make.bat``
file :-(

tk0miya commented:
LGTM. Confirmed using MacTeX.
It's okay to merge if @shimizukawa says ok.

shimizukawa reviewed (approved):
LGTM too.

jfbu committed:
Let clean target also remove xdv files

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -2,8 +2,15 @@
 ALLPDF = $(addsuffix .pdf,$(ALLDOCS))
+{% if latex_engine == 'xelatex' -%}
+ALLDVI =
+ALLXDV = $(addsuffix .xdv,$(ALLDOCS))
```
I have added this again, but only for "clean" target.

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -12,29 +19,52 @@ LATEXOPTS =
+
+{% if latex_engine == 'platex' -%}
+%.dvi: %.tex $(ALLIMGS) FORCE_MAKE
+	for f in *.pdf; do extractbb "$$f"; done
```
I have added ``"`` quotes. The other image types are already quoted for spaces in recipe. But I know Sphinx docs says somewhere filenames with spaces are not supported for images. Nevertheless adding the ``"`` for symmetry reasons.

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -12,29 +19,52 @@ LATEXOPTS =
+
+{% if latex_engine == 'platex' -%}
+%.pdf: %.tex $(ALLIMGS) FORCE_MAKE
+	for f in *.pdf; do extractbb "$$f"; done
```
Here too ``"$$f"`` rather than ``$$f``.

jfbu commented on sphinx/texinputs/Makefile_t:
```diff
@@ -57,30 +87,8 @@ bz2: tar
-
 clean:
-	rm -f *.log *.ind *.aux *.toc *.syn *.idx *.out *.ilg *.pla *.ps *.tar *.tar.gz *.tar.bz2 *.tar.xz $(ALLPDF) $(ALLDVI)
+	rm -f *.log *.ind *.aux *.toc *.syn *.idx *.out *.ilg *.pla *.ps *.tar *.tar.gz *.tar.bz2 *.tar.xz $(ALLPDF) $(ALLDVI) $(ALLXDV) *.fls *.fdb_latexmk
```
I do not use ``$(ALLDVI)$(ALLXDV)`` which avoids superfluous space because it would make mandatory that always one of them is empty, possibly creating error in future maintenance of this. For the record `.xdv` files are created by `xelatex` runs; if `xelatex` is run manually, the `xdv` file is removed but here it is run by Latexmk rule deliberately to not produced `pdf` until last final `xdvipdfmx`, for efficiency. So the `xdv` file remains in build repertory. (like `dvi` file when doing `platex+dvipdfmx`).

Side remark on legacy situation: this ``rm`` clearly may have unintended effects if ``$(ALLPDF)`` expands to filenames with spaces. Brief investigation on the net teaches me GNU Make at least is very awkward with spaces in filenames. Hence, Sphinx does not support here spaces in (basenames of) filenames of base documents. The shell globbing does not have this issue, but I think names coming from the Make variable have it. Anyway, this ``clean`` is for Users who actually ``cd`` to build repertory, which is not needed thanks to ``make latexpdf`` at top level, and ``make clean`` there is something else.

jfbu reviewed (commented):
@shimizukawa @tk0miya Thanks for all the time you devoted for reviewing multiple times. I have done a final polishing explained in comments, and if no objection I will merge this in 24hours. Thanks again !

jfbu committed:
Update latex builder docs (latexmk requirement)

jfbu commented:
Now merging :tada: thanks @shimizukawa @tk0miya for multiple reviews over the weeks.

@tk0miya how is this to be registered in CHANGES ? in ``Features added`` or in ``Incompatible changes``: it adds latexmk requirement to make latexpdf on unix-like platforms.

tk0miya commented:
>@tk0miya how is this to be registered in CHANGES ? in Features added or in Incompatible changes: it adds latexmk requirement to make latexpdf on unix-like platforms.

Let's put it on "incompatible changes" group.
And, please add new "dependencies" group and add latexmk to there.

jfbu commented:
ok thanks, will do. A bit worried that you never sleep, though ;-)

jfbu commented:
Done at 0f8abd1 !

tk0miya commented:
Thanks a lot!

jfbu commented:
@tk0miya regarding packaging for future 1.6 release, can you think about adding latexmk requirement to Unix platforms ? I don't know how you do that but I feel it might be mandatory; perhaps for example for good functioning of [readthedocs](https://readthedocs.org/). But I don't know if readthedocs pdf builds use the ``make latexpdf``. I know how to add Travis requirements, and currently we have not done that for latexmk, because Travis testing does only direct LaTeX runs and not via the ``make latexpdf``. As a result we have no testing for use of `latexmk`. And I fear the Python packaging might pick up requirements from the Travis ones.

tk0miya commented:
Sorry for late response.

>regarding packaging for future 1.6 release, can you think about adding latexmk requirement to Unix platforms ?

Yes, this issues added latexmk as a new dependencies.
I understand and agree that.
Let's see how it goes during beta release.

> we have not done that for latexmk, because Travis testing does only direct LaTeX runs and not via the make latexpdf. As a result we have no testing for use of latexmk. 

Right. It would be better to use latexmk on our testings.
